### PR TITLE
Add packet retransmission on timeout

### LINKED ISSUE #6: Add support for timeouts
- [ ] Client should give up after not hearing from the server after a duration has elapsed
- [ ] Server should attempt to resend packets after not receiving an ack for the last one it sent after a certain time period has elapsed
- [ ] Server should give up on resending after n resent packets

You can choose the value for 'n' and/or the timeout duration.

### PR DESCRIPTION
**Pull request by PurpleMyst**
Resolves #6

### CODE REVIEW

**Comment by connorkuehl** [comment]
Just wanted to let you know that this is on my radar. It's a bit of a larger PR so it might take me a little while longer to wrap my head around the changes and compile any feedback I have. Thanks! :slightly_smiling_face:

**Comment by PurpleMyst** [comment]
Alright, I appreciate it. Take your time, I understand it's an hefty PR :)

File: src/connection.rs [hunk 1]
```diff
@@ -148,4 +249,163 @@ mod tests {
...
+        // Then make sure our buffer contains the expected data and nothing weird happened
+        let buf = client_thread.join().unwrap().unwrap();
+        let mut expected = b"h".repeat(MAX_PAYLOAD_SIZE);
+        expected.push(b'i');
+        assert_eq!(buf, expected);
+    }
+
+    #[test]
+    fn test_put_retransmits_data() {
+        const BOGUS_DATA: &[u8] = b"hey, look, listen";
```
Comment (connorkuehl) [inline src/connection.rs:299]: Ha! 😀

File: src/server.rs [hunk 2]
```diff
@@ -12,21 +12,28 @@ use crate::bytes::{FromBytes, IntoBytes};
...
 
 impl Server {
     /// Creates a server configured to serve files from a given directory on
     /// a given address.
-    pub fn new<A: ToSocketAddrs, P: AsRef<Path>>(bind_to: A, serve_from: P) -> Result<Self> {
+    pub fn new<A: ToSocketAddrs, P: AsRef<Path>>(
+        bind_to: A,
+        serve_from: P,
+        retransmission_config: Option<RetransmissionConfig>,
+    ) -> Result<Self> {
```
Comment (connorkuehl) [inline src/server.rs:31]:
```suggestion
    pub fn new<A: ToSocketAddrs, P: AsRef<Path>>(
        bind_to: A,
        serve_from: P,
    ) -> Result<Self> {
```

Similar feedback to `Client`. Once `RetransmissionConfig` implements `Default`, let's just have this constructor use its default value.

File: src/server.rs [hunk 3]
```diff
@@ -36,12 +43,13 @@ impl Server {
     pub fn random_port<A: AsRef<str>, P: AsRef<Path>>(
         ip_addr: A,
         serve_from: P,
+        retransmission_config: Option<RetransmissionConfig>,
```
Comment (connorkuehl) [inline src/server.rs:46]:
```suggestion
    pub fn random_port<A: AsRef<str>, P: AsRef<Path>>(
        ip_addr: A,
        serve_from: P,
```

Same feedback as above

File: src/server.rs [hunk 4]
```diff
@@ -36,12 +43,13 @@ impl Server {
...
     ) -> Result<(u16, Self)> {
         let mut rng = rand::thread_rng();
         let port: u16 = rng.gen_range(MIN_PORT_NUMBER, u16::MAX);
         let bind_to = format!("{}:{}", ip_addr.as_ref(), port);
 
-        Self::new(bind_to, serve_from).map(|server| (port, server))
+        Self::new(bind_to, serve_from, retransmission_config).map(|server| (port, server))
```
Comment (connorkuehl) [inline src/server.rs:52]: Let's add a new method on `Server` 

`fn with_transmission_config(&mut self, cfg: RetransmissionConfig)`

and have it replace the default `RetransmissionConfig` it was constructed with

File: src/lib.rs [hunk 5]
```diff
@@ -39,6 +39,23 @@
 
 #![deny(missing_docs)]
 
+/// POD struct representing the configuration of the retransmission of packets
+// NB: this is a struct so that you can only specify max_retransmissions if you specify a time :>
```
Comment (connorkuehl) [inline src/lib.rs:43]: Reword this to be less technical (I.e., move away from phrases like "POD struct"

Maybe something like "Describes the retransmission behavior for a TFTP connection"

**Review by connorkuehl — CHANGES_REQUESTED** [review]
Really fantastic work, @PurpleMyst! 😀 

I have comments inline, but a summary:

1. Move `RetransmissionConfig` into the `connection` module (the `connection` module will have to be made `pub` iirc) 
1. Implement `Default` for RetransmissionConfig. This way we can get rid of all the `Option` stuff around the retransmission config as well as stop having to pass an `Option` RetransmissionConfig in type constructors.
    1. Construct the Server with the default config, and add a method on `Server` to let callers replace the default config.
    1. `client::{Builder, Config}` are a similar story, but update the Builder<ConnectTo> to have a default RetransmissionConfig and give it a method for callers to overwrite with their own config.
1. Reword the doc comment for `RetransmissionConfig`.

**Comment by PurpleMyst** [comment]
What should the default RetransmissionConfig do? Currently passing None turns on Retransmission indefinitely.

**Comment by connorkuehl** [comment]
> What should the default RetransmissionConfig do? Currently passing None turns on Retransmission indefinitely.

It can behave the same with an internal enum

```rust
enum Retransmit {
    Infinite,
    Finite(u64),
}
```

Or something else of your devising if you find something you like better!

**Comment by PurpleMyst** [comment]
Alright, I'll work on that. Are you sure that you want this in the `connection` module? We don't expose that currently, and exposing it just for `RetransmissionConfig` feels weird to me

**Review by connorkuehl — CHANGES_REQUESTED** [review]
I'd like to investigate ways to move more of the retransmission logic into types. The `check_retransmissions` method has a number of side effects in addition to checking if we should retransmit. These side effects include updating the `current_transmissions` variable as well as possibly writing to the socket.

Suppose we had an enum like this:

```rust
pub enum Retransmit {
  /// Do not retransmit packets. Just error out.
    Nil,
    /// Retransmit packets a limited amount of times
    Finite {
        /// How many times should we retransmit?
        limit: usize,
    },
    Infinite,
}
```

It could `impl Iterator` and depending on the variant, it could help control the retransmissions. The implementation of the `next()` method for this trait means that:

If it's `Nil`, it can just return `None`.

If it's `Finite` it can decrement its limit and return `Some(count)` (the actual value is insignificant, so long as it is a `Some` and not a `None` otherwise iteration would halt)

And if it's `Infinite` then it will just always return `Some`.

Then, the retransmission loops can be written as:

```rust
for _ in self.retransmission_config.retransmit.iter_mut() {
// if successful, break out, blah blah
}
```

Or something similar depending on how one will access the contents of the `RetransmissionConfig`. I think this will help dispense with the `check_retransmission` method and allow us to manage retransmissions without being so aware of different bits of state.

Thoughts?

**Comment by PurpleMyst** [comment]
First off, I'm not sure what the difference between my enum representation and yours is once you add back in the timeout `Duration`.

Secondly, I'm not sure I like the idea of using an `Iterator` simply for the fact that it makes the "send error packet on failure" case weird to handle. We'd need to break an enclosing labeled `loop`, put this into its own function that directly uses the `next` method, or use a `loop` with an `if let` and a `break`.

The `check_retransmissions` method is fine as is IMO, if you see it as less a pure check (one that would simply return a boolean) and more as an action and/or guard that is part of the retransmission process.

File: src/client.rs [hunk 6]
```diff
@@ -11,30 +11,31 @@ use crate::bytes::{FromBytes, IntoBytes};
 use crate::connection::Connection;
 use crate::connection::MIN_PORT_NUMBER;
 use crate::packet::*;
+use crate::RetransmissionConfig;
 
 /// The initial state for building a `Client`.
-pub struct New {
-    socket: UdpSocket,
-}
+pub struct New(());
```
[the remark this answers is not in the document]
Comment (PurpleMyst) [reply src/client.rs:17]: So that we could set the `RetransmissionConfig` at any time in the `Builder`.

Also, the reason there's one private field is to avoid people outside the crate constructing the `New` struct

**Comment by connorkuehl** [comment]
> First off, I'm not sure what the difference between my enum representation and yours is once you add back in the timeout `Duration`.

The suggested struct + enum representation doesn't duplicate the `Duration` across variant definitions.

> Secondly, I'm not sure I like the idea of using an `Iterator` simply for the fact that it makes the "send error packet on failure" case weird to handle. We'd need to break an enclosing labeled `loop`, put this into its own function that directly uses the `next` method, or use a `loop` with an `if let` and a `break`.

I was just thinking out loud when I said `Iterator`. One of my goals for this project is to practice more of "thinking in types" in the hopes that it makes it easier to create these logical flows we have to deal with.

> The `check_retransmissions` method is fine as is IMO, if you see it as less a pure check (one that would simply return a boolean) and more as an action and/or guard that is part of the retransmission process.

Have you experimented with an approach with fewer side effects?

**Comment by PurpleMyst** [comment]
> The suggested struct + enum representation doesn't duplicate the `Duration` across variant definitions.

That's valid, but IMHO it doesn't matter that much because 
1) It's laid out that way anyway by the rust compiler
2) It's more ergonomic for users this way

> Have you experimented with an approach with fewer side effects?

I could move the incrementation of check_retransmisions out of the function itself, but we'd still need to increment it somewhere and remember to do so anytime we used the function. This way, it's baked in.
